datafusion/catalog/src/table.rs

```diff
@@ -171,6 +172,34 @@ pub trait TableProvider: Debug + Sync + Send {
         limit: Option<usize>,
     ) -> Result<Arc<dyn ExecutionPlan>>;
 
+    async fn scan_with_args(
```

Comment:
I think this is a great new API and makes a lot of sense as it allows us to iterate / expand the scan API over time with less API disruption

It is also consistent with [`ScalarUDFImpl::invoke_with_args`](https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.ScalarUDFImpl.html#tymethod.invoke_with_args)

To merge this PR I think we should:
1. document this function (perhaps move the docs from `scan` here and then direct people to use `scan_with_args` with new TableProviders
2. Migrate all existing code in Datafusion to use `scan_with_args`
3. Consider deprecating `scan` 

(maybe we can do this as a follow on PR)